opm/simulators/linalg/bda/fgpilu.cpp

```diff
@@ -0,0 +1,570 @@
+            temp += b[block_size * row + k] * c[block_size * k + col];
+        }
+        a[block_size * row + col] -= temp;
+    }
+}
+
+
+void blockMult(
+    __local const double * restrict mat1,
+    __local const double * restrict mat2,
+    __global double * restrict resMat)
+{
+    const unsigned int block_size = 3;
+    const unsigned int warp_size = 32;
+    const unsigned int idx_t = get_local_id(0);                   // thread id in work group
+    const unsigned int thread_id_in_warp = idx_t % warp_size;     // thread id in warp (32 threads)
+    if(thread_id_in_warp < block_size * block_size){
+        const unsigned int row = thread_id_in_warp / block_size;
+        const unsigned int col = thread_id_in_warp % block_size;
+        double temp = 0.0;
+        for (unsigned int k = 0; k < block_size; k++) {
+            temp += mat1[block_size * row + k] * mat2[block_size * k + col];
+        }
+        resMat[block_size * row + col] = temp;
+    }
+}
+
+
+void invert(
+    __global const double * restrict matrix,
+    __local double * restrict inverse)
+{
+    const unsigned int block_size = 3;
+    const unsigned int bs = block_size;
+    const unsigned int warp_size = 32;
+    const unsigned int idx_t = get_local_id(0);                   // thread id in work group
+    const unsigned int thread_id_in_warp = idx_t % warp_size;     // thread id in warp (32 threads)
+    if(thread_id_in_warp < block_size * block_size){
+        // code generated by maple, copied from Dune::DenseMatrix
+        double t4  = matrix[0] * matrix[4];
+        double t6  = matrix[0] * matrix[5];
+        double t8  = matrix[1] * matrix[3];
+        double t10 = matrix[2] * matrix[3];
+        double t12 = matrix[1] * matrix[6];
+        double t14 = matrix[2] * matrix[6];
+
+        double det = (t4 * matrix[8] - t6 * matrix[7] - t8 * matrix[8] +
+                      t10 * matrix[7] + t12 * matrix[5] - t14 * matrix[4]);
+        double t17 = 1.0 / det;
+
+        const unsigned int r = thread_id_in_warp / block_size;
+        const unsigned int c = thread_id_in_warp % block_size;
+        const unsigned int r1 = (r+1) % bs;
+        const unsigned int c1 = (c+1) % bs;
+        const unsigned int r2 = (r+bs-1) % bs;
+        const unsigned int c2 = (c+bs-1) % bs;
+        inverse[c*bs+r] = ((matrix[r1*bs+c1] * matrix[r2*bs+c2]) - (matrix[r1*bs+c2] * matrix[r2*bs+c1])) * t17;
+    }
+}
+
+__kernel void fgpilu_sweep(
+    __global const double * restrict Ut_vals,
+    __global const double * restrict L_vals,
+    __global const double * restrict LU_vals,
+    __global const int * restrict Ut_rows,
+    __global const int * restrict L_rows,
+    __global const int * restrict LU_rows,
+    __global const int * restrict Ut_cols,
+    __global const int * restrict L_cols,
+    __global const int * restrict LU_cols,
+    __global double * restrict Ltmp,
+    __global double * restrict Utmp,
+    const int Nb,
```

Comment:
`Nb` means the number of blockrows of the matrix. I think `num_rows` is ambiguous, because it can mean both blocked or unblocked rows. Would `num_blockrows` be a good name?

Reply:
Nervermind.